File: open-nucleus-app/lib/core/constants/fhir_codes.dart
/// Common FHIR code systems and values used throughout the app.
///
/// These mirror the LOINC, encounter-status, condition-clinical, and
/// allergy-intolerance value sets that the Open Nucleus backend validates.
class FhirCodes {
  FhirCodes._();

  // ---------------------------------------------------------------------------
  // LOINC vital-sign codes
  // ---------------------------------------------------------------------------

  static const String loincSystem = 'http://loinc.org';

  static const String heartRate = '8867-4';
  static const String bloodPressureSystolic = '8480-6';
  static const String bloodPressureDiastolic = '8462-4';
  static const String bodyTemperature = '8310-5';
  static const String respiratoryRate = '9279-1';
  static const String oxygenSaturation = '2708-6';
  static const String bodyWeight = '29463-7';
  static const String bodyHeight = '8302-2';
  static const String bmi = '39156-5';

  /// Human-readable display strings keyed by LOINC code.
  static const Map<String, String> vitalDisplayNames = {
    heartRate: 'Heart Rate',
    bloodPressureSystolic: 'Systolic Blood Pressure',
    bloodPressureDiastolic: 'Diastolic Blood Pressure',
    bodyTemperature: 'Body Temperature',
    respiratoryRate: 'Respiratory Rate',
    oxygenSaturation: 'SpO2',
    bodyWeight: 'Weight',
    bodyHeight: 'Height',
    bmi: 'BMI',
  };

  /// Standard units for each vital sign.
  static const Map<String, String> vitalUnits = {
    heartRate: 'bpm',
    bloodPressureSystolic: 'mmHg',
    bloodPressureDiastolic: 'mmHg',
    bodyTemperature: 'Cel',
    respiratoryRate: '/min',
    oxygenSaturation: '%',
    bodyWeight: 'kg',
    bodyHeight: 'cm',
    bmi: 'kg/m2',
  };

  // ---------------------------------------------------------------------------
  // Encounter status (http://hl7.org/fhir/encounter-status)
  // ---------------------------------------------------------------------------

  static const String encounterStatusSystem =
      'http://hl7.org/fhir/encounter-status';

  static const String encounterPlanned = 'planned';
  static const String encounterArrived = 'arrived';
  static const String encounterTriaged = 'triaged';
  static const String encounterInProgress = 'in-progress';
  static const String encounterOnLeave = 'onleave';
  static const String encounterFinished = 'finished';
  static const String encounterCancelled = 'cancelled';

  static const List<String> encounterStatuses = [
    encounterPlanned,
    encounterArrived,
    encounterTriaged,
    encounterInProgress,
    encounterOnLeave,
    encounterFinished,
    encounterCancelled,
  ];

  // ---------------------------------------------------------------------------
  // Condition clinical status
  // (http://terminology.hl7.org/CodeSystem/condition-clinical)
  // ---------------------------------------------------------------------------

  static const String conditionClinicalSystem =
      'http://terminology.hl7.org/CodeSystem/condition-clinical';

  static const String conditionActive = 'active';
  static const String conditionRecurrence = 'recurrence';
  static const String conditionRelapse = 'relapse';
  static const String conditionInactive = 'inactive';
  static const String conditionRemission = 'remission';
  static const String conditionResolved = 'resolved';

  static const List<String> conditionClinicalStatuses = [
    conditionActive,
    conditionRecurrence,
    conditionRelapse,
    conditionInactive,
    conditionRemission,
    conditionResolved,
  ];

  // ---------------------------------------------------------------------------
  // AllergyIntolerance criticality
  // (http://hl7.org/fhir/allergy-intolerance-criticality)
  // ---------------------------------------------------------------------------

  static const String allergyCriticalitySystem =
      'http://hl7.org/fhir/allergy-intolerance-criticality';

  static const String criticalityLow = 'low';
  static const String criticalityHigh = 'high';
  static const String criticalityUnableToAssess = 'unable-to-assess';

  static const List<String> allergyCriticalities = [
    criticalityLow,
    criticalityHigh,
    criticalityUnableToAssess,
  ];
}

File: open-nucleus-app/src/lib/fhir-codes.ts
/** A vital sign definition with its LOINC code, display name, and default unit. */
export interface VitalSignDef {
  code: string;
  display: string;
  unit: string;
}

/** Common vital sign LOINC codes used in observation forms. */
export const VITAL_SIGNS: VitalSignDef[] = [
  { code: '8310-5', display: 'Body Temperature', unit: 'degC' },
  { code: '8867-4', display: 'Heart Rate', unit: '/min' },
  { code: '85354-9', display: 'Blood Pressure Panel', unit: 'mmHg' },
  { code: '8480-6', display: 'Systolic Blood Pressure', unit: 'mmHg' },
  { code: '8462-4', display: 'Diastolic Blood Pressure', unit: 'mmHg' },
  { code: '29463-7', display: 'Body Weight', unit: 'kg' },
  { code: '8302-2', display: 'Body Height', unit: 'cm' },
  { code: '2708-6', display: 'Oxygen Saturation (SpO2)', unit: '%' },
  { code: '9279-1', display: 'Respiratory Rate', unit: '/min' },
];

/** FHIR Encounter status codes. */
export const ENCOUNTER_STATUSES = [
  'planned',
  'arrived',
  'triaged',
  'in-progress',
  'onleave',
  'finished',
  'cancelled',
] as const;

/** FHIR Condition clinical status codes. */
export const CONDITION_STATUSES = [
  'active',
  'recurrence',
  'relapse',
  'inactive',
  'remission',
  'resolved',
] as const;

/** FHIR AllergyIntolerance criticality codes. */
export const ALLERGY_CRITICALITIES = [
  'low',
  'high',
  'unable-to-assess',
] as const;

/** FHIR MedicationRequest status codes. */
export const MEDICATION_STATUSES = [
  'active',
  'on-hold',
  'cancelled',
  'completed',
  'entered-in-error',
  'stopped',
  'draft',
  'unknown',
] as const;

/** FHIR Immunization status codes. */
export const IMMUNIZATION_STATUSES = [
  'completed',
  'entered-in-error',
  'not-done',
] as const;

/** FHIR Procedure status codes. */
export const PROCEDURE_STATUSES = [
  'preparation',
  'in-progress',
  'not-done',
  'on-hold',
  'stopped',
  'completed',
  'entered-in-error',
  'unknown',
] as const;

/** FHIR Observation status codes. */
export const OBSERVATION_STATUSES = [
  'registered',
  'preliminary',
  'final',
  'amended',
  'corrected',
  'cancelled',
  'entered-in-error',
  'unknown',
] as const;

/** FHIR administrative gender codes. */
export const ADMINISTRATIVE_GENDERS = [
  'male',
  'female',
  'other',
  'unknown',
] as const;

/** FHIR AllergyIntolerance category codes. */
export const ALLERGY_CATEGORIES = [
  'food',
  'medication',
  'environment',
  'biologic',
] as const;

/** FHIR AllergyIntolerance type codes. */
export const ALLERGY_TYPES = [
  'allergy',
  'intolerance',
] as const;

/** FHIR Consent status codes. */
export const CONSENT_STATUSES = [
  'draft',
  'proposed',
  'active',
  'rejected',
  'inactive',
  'entered-in-error',
] as const;
